File: core/src/main/java/ru/pocketbyte/locolaser/resource/entity/ResMap.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.resource.entity;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Denis Shurygin
 */
public class ResMap extends HashMap<String, ResLocale> {

    public ResMap() {
        super();
    }

    public ResMap(Map<String, ResLocale> map) {
        super();
        if (map != null)
            for (Entry<String, ResLocale> entry: map.entrySet()) {
                put(entry.getKey(), new ResLocale(entry.getValue()));
            }
    }

    public ResMap merge(ResMap map) {
        if (map != null) {
            for (Map.Entry<String, ResLocale> itemsEntry : map.entrySet()) {
                String locale = itemsEntry.getKey();

                ResLocale destinationItems = this.get(locale);
                if (destinationItems != null)
                    destinationItems = destinationItems.merge(itemsEntry.getValue());
                else
                    destinationItems = itemsEntry.getValue();

                this.put(locale, destinationItems);
            }
        }
        return this;
    }

    public ResMap remove(ResMap map) {
        if (map != null) {
            for (Map.Entry<String, ResLocale> itemsEntry : map.entrySet()) {
                String locale = itemsEntry.getKey();
                ResLocale removeItem = itemsEntry.getValue();
                ResLocale destinationItems = this.get(locale);

                if (destinationItems != null) {
                    destinationItems.remove(removeItem);
                    if (destinationItems.size() == 0)
                        this.remove(locale);
                }
            }
        }
        return this;
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/resource/entity/ResMap.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.resource.entity

import ru.pocketbyte.locolaser.config.resources.filter.ResourcesFilter
import java.util.HashMap

/**
 * @author Denis Shurygin
 */
class ResMap() : HashMap<String, ResLocale>() {

    constructor(map: Map<String, ResLocale>?) : this() {
        map?.forEach { key, value ->
            put(key, ResLocale(value))
        }
    }

    fun remove(map: ResMap?): ResMap {
        map?.forEach { locale, removeItem ->
            val destinationItems = this[locale]

            destinationItems?.remove(removeItem)
            if (destinationItems?.size == 0)
                this.remove(locale)
        }
        return this
    }
}

fun ResMap?.merge(map: ResMap?): ResMap? {
    if (this == null)
        return map

    map?.forEach { locale, value ->
        this[locale] = this[locale]?.merge(value) ?: value
    }
    return this
}

inline fun ResMap.filter(filter: ((key: String) -> Boolean)): ResMap {
    val newResMap = ResMap()

    this.forEach { (key, resLocale) ->
        newResMap[key] = resLocale.filter(filter)
    }

    return newResMap
}

fun ResMap.filter(filter: ResourcesFilter?): ResMap {
    if (filter == null) {
        return this
    }

    return filter { key ->
        filter.filter(key)
    }
}